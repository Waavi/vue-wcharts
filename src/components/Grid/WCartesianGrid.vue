<template>
    <g>
        <line
            v-for="(line, index) in lines"
            :key="`gridLine${index}`"
            :x1="line.x1"
            :y1="line.y1"
            :x2="line.x2"
            :y2="line.y2"
            :stroke="stylesCmp.stroke"
            :stroke-width="stylesCmp.strokeWidth"
            :stroke-dasharray="stylesCmp.strokeDasharray"
            fill="none"
        />
    </g>
</template>

<script>
import VueTypes from 'vue-types'
import { genTicks, genExactNbTicks } from '../../utils/maths'

const stylesDefaultProp = {
    stroke: '#ccc',
    strokeWidth: 1,
    strokeDasharray: '3',
}

export default {
    name: 'WCartesianGrid',
    type: 'cartesian',
    inject: ['Chart'],
    props: {
        hideH: VueTypes.bool.def(false),
        hideV: VueTypes.bool.def(false),
        numLinesH: VueTypes.number.def(0),
        numLinesV: VueTypes.number.def(0),
        styles: VueTypes.shape({
            stroke: VueTypes.string,
            strokeWidth: VueTypes.number,
            strokeDasharray: VueTypes.oneOfType([String, Number]),
        }).def(() => ({
            ...stylesDefaultProp,
        })),
    },
    computed: {
        stylesCmp () {
            return {
                ...stylesDefaultProp,
                ...this.styles,
            }
        },
        hLines () {
            // Return a empty array if we don't want to show horizontal lines
            if (this.hideH) return []
            const {
                dataset, canvas, bounds, yScale,
            } = this.Chart
            // Calculate number of lines to paint
            const numLines = this.numLinesH || dataset.length
            // Select the correct function and generate the value of the lines. ex: [100, 500, 1500, 2500, 5000]
            const getLinesFn = this.numLinesH ? genExactNbTicks : genTicks
            const lines = getLinesFn(bounds.min, bounds.max, numLines).reverse()
            // Generate lines coordinates
            return lines.map((value, index) => {
                const y = yScale(value)
                return {
                    x1: canvas.x0,
                    y1: y,
                    x2: canvas.x1,
                    y2: y,
                }
            })
        },
        vLines () {
            // Return a empty array if we don't want to show vertical lines
            if (this.hideV) return []
            const {
                dataset, canvas, padding, xBounds, xScale, scatter,
            } = this.Chart
            // Calculate number of lines to generate
            const numLines = this.numLinesV || dataset.length
            if (scatter) {
                // Select the correct function and generate the value of the lines. ex: [100, 500, 1500, 2500, 5000]
                const getLinesFn = this.numLinesH ? genExactNbTicks : genTicks
                const lines = getLinesFn(xBounds.min, xBounds.max, numLines).reverse()
                // Generate lines coordinates
                return lines.map((value, index) => {
                    const x = xScale(value)
                    return {
                        x1: x,
                        y1: canvas.y0,
                        x2: x,
                        y2: canvas.y1,
                    }
                })
            }
            // Calculate left and right space
            const offset = padding[1] + padding[3]
            // Calculate space between lines
            const space = (canvas.width - offset) / (numLines - 1)
            // Generate lines coordinates

            return Array.from({ length: numLines }).map((props, index) => {
                // X position is calculated taking x0 canvas position adding space between lines times index (0...numLines-1) and adding left space
                const x = canvas.x0 + (space * index) + padding[3]
                return {
                    x1: x,
                    y1: canvas.y0,
                    x2: x,
                    y2: canvas.y1,
                }
            })
        },
        lines () {
            return [...this.hLines, ...this.vLines]
        },
    },
}
</script>
