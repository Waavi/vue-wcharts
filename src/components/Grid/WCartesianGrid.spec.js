import { shallowMount, mount } from '@vue/test-utils'
import WCartesianGrid from './WCartesianGrid.vue'

describe('Components/WCartesianGrid', () => {
    const dataset = [
        {
            name: 'Page A', one: 4000, two: -2400, three: 2300, four: 1200, five: 2300,
        },
        {
            name: 'Page B', one: 3000, two: 2210, three: 1398, four: 1200, five: 1398,
        },
        {
            name: 'Page C', one: 9800, two: 2200, three: 0, four: 1200, five: 0,
        },
    ]

    const propsData = {
        hideH: true,
        hideV: true,
    }

    const provide = {
        Chart: {
            dataset,
            data: dataset,
        },
    }

    const propsDataWithLines = {
        hideH: false,
        hideV: false,
        numLinesH: 1,
        numLinesV: 1,
        styles: {
            stroke: '000',
            strokeWidth: 1,
            strokeDasharray: 0,
        },
    }

    const provideWithLines = {
        ...provide,
        Chart: {
            ...provide.Chart,
            dataset,
            data: dataset,
            canvas: {
                width: 680,
                x0: 50,
                y0: 25,
                x1: 500,
                y1: 250,
            },
            padding: [0, 0, 0, 0],
            bounds: {
                max: 9490,
                min: 780,
            },
            xBounds: {
                max: 9490,
                min: 780,
            },
            xScale: a => a,
            yScale: a => a,
        },
    }

    const defaultConfig = {
        propsData,
        provide,
    }

    const defaultConfigWithLines = {
        propsData: propsDataWithLines,
        provide: provideWithLines,
    }

    it(`Should be render correctly without lines`, () => {
        const wrapper = shallowMount(WCartesianGrid, defaultConfig)
        expect(wrapper.html()).toMatchSnapshot()
    })

    it(`Should be render correctly with lines`, () => {
        const wrapper = mount(WCartesianGrid, defaultConfigWithLines)
        expect(wrapper.html()).toMatchSnapshot()
    })

    it(`Should be render correctly with 3 lines`, () => {
        const wrapper = mount(WCartesianGrid, defaultConfigWithLines)
        expect(wrapper.findAll('line')).toHaveLength(3)
    })
})
