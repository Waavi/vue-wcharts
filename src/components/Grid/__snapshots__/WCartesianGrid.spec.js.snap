// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/WLine Should be render correctly with lines 1`] = `
<g>
  <line x1="50" y1="9490" x2="500" y2="9490" stroke="000" stroke-width="1" stroke-dasharray="0" fill="none"></line>
  <line x1="50" y1="780" x2="500" y2="780" stroke="000" stroke-width="1" stroke-dasharray="0" fill="none"></line>
  <line x1="NaN" y1="25" x2="NaN" y2="250" stroke="000" stroke-width="1" stroke-dasharray="0" fill="none"></line>
</g>
`;

exports[`Components/WLine Should be render correctly without lines 1`] = `<g></g>`;
